File: saizad-mvvm/src/main/java/com/saizad/mvvm/delegation/BaseCB.java
package com.saizad.mvvm.delegation;

import android.content.Context;

import androidx.annotation.MenuRes;
import androidx.annotation.NonNull;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.ViewModelStoreOwner;
import androidx.navigation.NavController;

import com.saizad.mvvm.Environment;
import com.saizad.mvvm.SaizadLocation;
import com.saizad.mvvm.ViewModelProviderFactory;
import com.saizad.mvvm.components.SaizadBaseViewModel;

public interface BaseCB<V extends SaizadBaseViewModel> {
    @NonNull
    LifecycleOwner getViewLifecycleOwner();

    @NonNull
    Context context();

    @NonNull
    ViewModelStoreOwner viewModelStoreOwner();

    @NonNull
    ViewModelProviderFactory viewModelProviderFactory();

    @NonNull
    Class<V> getViewModelClassType();

    @NonNull
    SaizadLocation appLocation();

    @NonNull
    Environment environment();

    @MenuRes
    int menRes();

    NavController navController();
}

File: saizad-mvvm/src/main/java/com/saizad/mvvm/delegation/BaseCB.kt
package com.saizad.mvvm.delegation

import android.content.Context
import android.os.Bundle
import androidx.annotation.MenuRes
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.ViewModelStoreOwner
import androidx.navigation.NavController
import com.saizad.mvvm.SaizadLocation
import com.saizad.mvvm.components.SaizadBaseViewModel
import sa.zad.easypermission.PermissionManager

interface BaseCB<V : SaizadBaseViewModel> {

    val lifecycleOwner: LifecycleOwner
    fun context(): Context
    fun viewModelStoreOwner(): ViewModelStoreOwner
    val viewModelClassType: Class<V>
    fun appLocation(): SaizadLocation
    fun permissionManager(): PermissionManager

    @MenuRes
    fun menRes(): Int
    fun navController(): NavController
    fun bundle(): Bundle?
}